osquery/tables/system/windows/certificates.cpp

```diff
@@ -0,0 +1,300 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#define _WIN32_DCOM
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <Wintrust.h>
+#include <wincrypt.h>
+
+#include <boost/algorithm/hex.hpp>
+#include <boost/functional/hash/hash.hpp>
+
+#include <osquery/database.h>
+#include <osquery/logger.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/core/windows/wmi.h"
+#include "osquery/filesystem/fileops.h"
+
+namespace osquery {
+namespace tables {
+
+#define CERT_ENCODING (PKCS_7_ASN_ENCODING | X509_ASN_ENCODING)
+
+const std::map<unsigned long, std::string> kKeyUsages = {
+    {CERT_DATA_ENCIPHERMENT_KEY_USAGE, "CERT_DATA_ENCIPHERMENT_KEY_USAGE"},
+    {CERT_DIGITAL_SIGNATURE_KEY_USAGE, "CERT_DIGITAL_SIGNATURE_KEY_USAGE"},
+    {CERT_KEY_AGREEMENT_KEY_USAGE, "CERT_KEY_AGREEMENT_KEY_USAGE"},
+    {CERT_KEY_CERT_SIGN_KEY_USAGE, "CERT_KEY_CERT_SIGN_KEY_USAGE"},
+    {CERT_KEY_ENCIPHERMENT_KEY_USAGE, "CERT_KEY_ENCIPHERMENT_KEY_USAGE"},
+    {CERT_NON_REPUDIATION_KEY_USAGE, "CERT_NON_REPUDIATION_KEY_USAGE"},
+    {CERT_OFFLINE_CRL_SIGN_KEY_USAGE, "CERT_OFFLINE_CRL_SIGN_KEY_USAGE"}};
+
+typedef struct _ENUM_ARG {
+  DWORD dwFlags;
+  const void* pvStoreLocationPara;
+  QueryData* results;
+  std::set<std::string>* processed;
+} ENUM_ARG, *PENUM_ARG;
+
+std::string cryptOIDToString(const char* objId) {
+  if (objId == nullptr) {
+    return "";
+  }
+  auto objKeyId = const_cast<char*>(objId);
+  auto oidInfo = CryptFindOIDInfo(CRYPT_OID_INFO_OID_KEY, objKeyId, 0);
+  return oidInfo == nullptr ? "" : wstringToString(oidInfo->pwszName);
+}
+
+std::string getKeyUsage(const PCERT_INFO& certInfo) {
+  // Key usage size is 1 or 2 bytes of data, we use 4 to cast to int
+  const size_t keyUsageSize = 4;
+  BYTE keyUsageBuff[keyUsageSize];
+  auto ret = CertGetIntendedKeyUsage(
+      CERT_ENCODING, certInfo, keyUsageBuff, keyUsageSize);
+  if (ret == 0) {
+    return "";
+  }
+  std::vector<std::string> usages;
+  auto usage = reinterpret_cast<size_t>(keyUsageBuff);
```

Comment:
Not sure if you should use `size_t`, do you want a bit array? If so choose some container that is always size=4, like a `uint32_t`.